sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregateCodegenSupport.scala

```diff
@@ -47,6 +47,12 @@ trait AggregateCodegenSupport
    */
   private var bufVars: Seq[Seq[ExprCode]] = _
 
+  /**
+   * Whether this operator has an aggregate build phase.
+   * Only [[HashAggregateExec]] has build phase now.
+   */
+  protected def hasAggBuild: Boolean
```

Comment:
shall we rename it to `needHashTable`? And it should be false in `HashAggregateExec` if group by col is empty.

Reply:
do we want to report `aggTime` as 0 now for `HashAggregate` without group by columns?

I am fine with changing to report 0 now, as the SQL metrics shown up as `time in aggregation build` in Spark UI for end users. There's no build for `HashAggregate` without group by columns, so I think it's a valid fix. Just want to mention here as this is a user-facing change.

Reply:
ok maybe we can do the change in a new PR. Let's just use a proper name in this PR.

Reply:
@cloud-fan - sure, changed to `needHashTable`. Leave `HashAggregateExec.needHashTable` to be true, to maintain existing behavior for now.